File: Frontend/src/service/news/newsService.js
import BaseService from '@/service/baseService'

/**
 * Service class for handling news-related API operations
 * @extends BaseService
 */
class NewsService extends BaseService {
  /**
   * Creates an instance of NewsService with the news API endpoint
   */
  constructor() {
    super('/news')
  }

  /**
   * Retrieves paginated news from the system
   * @async
   * @param {number} page - The page number to retrieve
   * @param {number} size - The number of items per page
   * @returns {Promise<Object>} Promise resolving to an object containing news array and pagination info
   */
  async fetchPaginatedNews(page, size) {
    const response = await this.get(`get/?page=${page}&size=${size}`)
    const data = response
    
    return {
      news: data.news || [],
      totalPages: data.totalPages || 1,
      totalElements: data.totalElements || 0,
    }
  }

  /**
   * Retrieves a specific news item by its ID
   * @async
   * @param {number} id - The unique identifier of the news to retrieve
   * @returns {Promise<Object>} Promise resolving to the news object
   */
  async getNewsById(id) {
    return await this.get(`${id}`)
  }

  /**
   * Creates a new news item in the system
   * @async
   * @param {Object} newsData - The data for the new news item
   * @returns {Promise<Object>} Promise resolving to the API response
   */
  async createNews(newsData) {
    return await this.post('create', newsData)
  }

  /**
   * Deletes a news item from the system
   * @async
   * @param {number} id - The unique identifier of the news item to delete
   * @returns {Promise<Object>} Promise resolving to the API response
   */
  async deleteNews(id) {
    return await this.post(`delete/${id}`)
  }

  /**
   * Updates an existing news item in the system
   * @async
   * @param {number} id - The unique identifier of the news item to update
   * @param {Object} newsData - The updated news data
   * @returns {Promise<Object>} Promise resolving to the API response
   */
  async updateNews(id, newsData) {
    return await this.post(`edit/${id}`, newsData)
  }
}

export default new NewsService()

File: Backend/src/main/java/edu/ntnu/idatt2106/krisefikser/service/news/NewsService.java
package edu.ntnu.idatt2106.krisefikser.service.news;

import edu.ntnu.idatt2106.krisefikser.api.dto.news.EditNewsDto;
import edu.ntnu.idatt2106.krisefikser.persistance.entity.news.News;
import edu.ntnu.idatt2106.krisefikser.persistance.repository.news.NewsRepository;
import java.time.LocalDateTime;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.stereotype.Service;

/**
 * The type News service.
 */
@Service
public class NewsService {

  private final NewsRepository newsRepository;

  /**
   * The NewsService class provides methods for managing news items.
   *
   * @param newsRepository the news repository
   */
  public NewsService(NewsRepository newsRepository) {
    this.newsRepository = newsRepository;
  }

  /**
   * Creates a new news item.
   *
   * @param newsDto the news item to create
   */
  public void createNewsItem(EditNewsDto newsDto) {
    News news = new News();
    news.setTitle(newsDto.getTitle());
    news.setUrl(newsDto.getUrl());
    news.setContent(newsDto.getContent());
    news.setSource(newsDto.getSource());
    news.setCreatedAt(LocalDateTime.now());

    newsRepository.save(news);
  }

  /**
   * Updates an existing news item.
   *
   * @param id      the ID of the news item to update
   * @param newsDto the updated news item data
   */
  public void updateNewsItem(Long id, EditNewsDto newsDto) {
    News news = newsRepository.findById(id)
        .orElseThrow(() -> new IllegalArgumentException("News with given id not found"));
    if (news != null) {
      if (newsDto.getTitle() != null) {
        news.setTitle(newsDto.getTitle());
      }
      if (newsDto.getUrl() != null) {
        news.setUrl(newsDto.getUrl());
      }
      if (newsDto.getContent() != null) {
        news.setContent(newsDto.getContent());
      }
      if (newsDto.getSource() != null) {
        news.setSource(newsDto.getSource());
      }

      newsRepository.save(news);
    } else {
      throw new IllegalArgumentException("News with id " + id + " not found");
    }
  }

  /**
   * Deletes a news item.
   *
   * @param id the ID of the news item to delete
   */
  public void deleteNewsItem(Long id) {
    News news = newsRepository.findById(id)
        .orElseThrow(() -> new IllegalArgumentException("News with given id not found"));
    if (news != null) {
      newsRepository.delete(news);
    } else {
      throw new IllegalArgumentException("News with id " + id + " not found");
    }
  }

  /**
   * Finds a news item by its ID.
   *
   * @param id the ID of the news item
   * @return the found news item, or null if not found
   */
  public News findNewsById(Long id) {
    return newsRepository.findById(id)
        .orElseThrow(() -> new IllegalArgumentException("News with given id not found"));
  }

  /**
   * Finds a paginated list of news items.
   *
   * @param page the current page number (0-indexed)
   * @param size the number of items per page
   * @return a paginated list of news items
   */
  public Page<News> findPaginatedList(int page, int size) {
    return newsRepository.findByOrderByCreatedAtDesc(PageRequest.of(page, size));
  }
}
